types/src/v0/impls/reward.rs

```diff
@@ -0,0 +1,496 @@
+    };
+
+    let computed_rewards = compute_rewards(validator)?;
+    for (address, reward) in computed_rewards {
+        update_balance(&RewardAccount(address.to_ethers()), reward)?;
+    }
+    Ok(reward_state)
+}
+
+pub fn compute_rewards(
+    validator: Validator<BLSPubKey>,
+) -> anyhow::Result<Vec<(alloy::primitives::Address, RewardAmount)>> {
+    ensure!(
+        validator.commission <= COMMISSION_BASIS_POINTS,
+        "commission must not exceed {COMMISSION_BASIS_POINTS}"
+    );
+
+    let mut rewards = Vec::new();
+
+    let total_reward = block_reward().0;
+    let delegators_ratio_basis_points = U256::from(COMMISSION_BASIS_POINTS)
+        .checked_sub(U256::from(validator.commission))
+        .context("overflow")?;
+    let delegators_reward = delegators_ratio_basis_points
+        .checked_mul(total_reward)
+        .context("overflow")?;
+
+    // Distribute delegator rewards
+    let total_stake = validator.stake.to_ethers();
+    let mut delegators_rewards_distributed = U256::from(0);
+    for (delegator_address, delegator_stake) in &validator.delegators {
+        let delegator_reward = RewardAmount::from(
+            (delegator_stake
+                .to_ethers()
+                .checked_mul(delegators_reward)
+                .context("overflow")?
+                .checked_div(total_stake)
+                .context("overflow")?)
+            .checked_div(COMMISSION_BASIS_POINTS.into())
+            .context("overflow")?,
+        );
+
+        delegators_rewards_distributed += delegator_reward.0;
+
+        rewards.push((*delegator_address, delegator_reward));
+    }
+
+    let leader_reward = total_reward
+        .checked_sub(delegators_rewards_distributed)
+        .context("overflow")?;
+    rewards.push((validator.account, leader_reward.into()));
+
+    Ok(rewards)
+}
+
+pub async fn catchup_missing_accounts(
+    instance_state: &NodeState,
+    validated_state: &mut ValidatedState,
+    parent_leaf: &Leaf2,
+    view: ViewNumber,
+) -> anyhow::Result<Validator<BLSPubKey>> {
+    let height = parent_leaf.height();
+    let epoch_height = instance_state.epoch_height.unwrap();
+    let epoch = EpochNumber::new(height % epoch_height);
+    let coordinator = instance_state.coordinator.clone();
+
+    let epoch_membership = coordinator.membership_for_epoch(Some(epoch)).await?;
+    let membership = epoch_membership.coordinator.membership().read().await;
+
+    let leader: BLSPubKey = membership
+        .leader(ViewNumber::new(height), Some(epoch))
+        .context(format!("leader for epoch {epoch:?} not found"))?;
+
+    let validator = membership.get_validator_config(&epoch, leader).unwrap();
+    let mut reward_accounts = vec![validator.account.to_ethers().into()];
+    let delegators = validator
+        .delegators
+        .keys()
+        .cloned()
+        .map(|a| a.to_ethers().into())
+        .collect::<Vec<RewardAccount>>();
+
+    reward_accounts.extend(delegators.clone());
```

Comment:
@imabdulbasit if the validator is also a delegator we will have it twice in the `reward_accounts`. It's fine I guess but we could use a set to avoid it.

Reply:
https://github.com/EspressoSystems/espresso-network/pull/2772/commits/ebe341b390c4222fe0d135a8065b2ee85f0da61a